thrift/arg2/kv_iterator.go

```diff
@@ -0,0 +1,88 @@
+// Package arg2 contains tchannel thrift Arg2 interfaces for external use.
+//
+// These interfaces are currently unstable, and aren't covered by the API
+// backwards-compatibility guarantee.
+package arg2
+
+import "encoding/binary"
+
+// KeyValIterator is a iterator for reading tchannel-thrift Arg2 Scheme,
+// which has key/value pairs (k~2 v~2).
+// NOTE: to be optimized for performance, we try to limit the allocation
+// done in the process of iteration.
+type KeyValIterator struct {
+	arg2Payload           []byte
+	leftPairCount         int
+	keyOffset             int
+	valueOffset, valueLen int
+}
+
+// InitKeyValIterator inits a KeyValIterator with the buffer pointing at
+// start of Arg2.
+// NOTE: tchannel-thrift Arg Scheme starts with number of key/value pair.
+func InitKeyValIterator(arg2Payload []byte) (KeyValIterator, bool) {
+	if len(arg2Payload) < 2 {
+		return KeyValIterator{}, false
+	}
+
+	nh := int(binary.BigEndian.Uint16(arg2Payload[0:2]))
+	if nh <= 0 {
```

Comment:
is this branch worth having? it seems like the resulting slice without this will be basically the same

Reply:
I think it's worthy having this branch as it won't cause much for the if check on an cached value, and it provides the opportunity to exit quit.

Also based on existing code, if the arg2 length is set to be long enough, but nh == 0 (invalid case), ignoring the branch could cause the iterator iterating on an invalid kv pair.

Reply:
Since we're using a `Uint16`, the value can never be negative -- it can only be 0 or positive right? The 0 value is positive that we handle already.

If we replace this branch, and do `.Next()`, then we're essentially doing the same check just in `Next` instead of here (we're creating a `KeyValueIterator`, but not sure that the optimization for that is worth it).

If we do that, we probably don't even have a need for `next(..)` since the logic can be inlined into `Next(.)`